deployment/src/main/java/com/vaadin/quarkus/deployment/VaadinQuarkusProcessor.java

```diff
@@ -15,29 +15,108 @@
 
 import io.quarkus.arc.deployment.BeanDefiningAnnotationBuildItem;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.undertow.deployment.ServletBuildItem;
+import org.jboss.jandex.IndexView;
 
 class VaadinQuarkusProcessor {
 
     private static final String FEATURE = "vaadin-quarkus";
 
-    private static final DotName ROUTE_ANNOTATION = DotName.createSimple(Route.class.getName());
+    private static final DotName ROUTE_ANNOTATION = DotName
+            .createSimple(Route.class.getName());
 
     @BuildStep
     FeatureBuildItem feature() {
         return new FeatureBuildItem(FEATURE);
     }
-    
+
     @BuildStep
-    public void build(final BuildProducer<BeanDefiningAnnotationBuildItem> additionalBeanDefiningAnnotationRegistry) {
+    public void build(
+            final BuildProducer<AdditionalBeanBuildItem> additionalBeanProducer,
+            final BuildProducer<BeanDefiningAnnotationBuildItem> additionalBeanDefiningAnnotationRegistry) {
+        additionalBeanProducer.produce(AdditionalBeanBuildItem
+                .unremovableOf(QuarkusVaadinServlet.class));
+
         // Make and Route annotated Component a bean for injection
-        additionalBeanDefiningAnnotationRegistry.produce(new BeanDefiningAnnotationBuildItem(ROUTE_ANNOTATION));
+        additionalBeanDefiningAnnotationRegistry
+                .produce(new BeanDefiningAnnotationBuildItem(ROUTE_ANNOTATION));
+    }
+
+    @BuildStep
+    void mapVaadinServletPaths(final BeanArchiveIndexBuildItem beanArchiveIndex,
+            final BuildProducer<ServletBuildItem> servletProducer) {
+        final IndexView indexView = beanArchiveIndex.getIndex();
+
+        // Collect all VaadinServlet instances and remove QuarkusVaadinServlet
+        // and VaadinServlet from the list.
+        final Collection<ClassInfo> vaadinServlets = indexView
+                .getAllKnownSubclasses(
+                        DotName.createSimple(VaadinServlet.class.getName()))
+                .stream().filter(servlet -> !servlet.name().toString()
+                        .equals(QuarkusVaadinServlet.class.getName()) || servlet
+                        .name().toString()
+                        .equals(VaadinServlet.class.getName()))
+                .collect(Collectors.toList());
+
+        // If no VaadinServlet instances found register QuarkusVaadinServlet
+        if (vaadinServlets.isEmpty()) {
+            servletProducer.produce(ServletBuildItem
+                    .builder(QuarkusVaadinServlet.class.getName(),
+                            QuarkusVaadinServlet.class.getName())
+                    .addMapping("/*").setAsyncSupported(true).build());
+        } else {
+            registerUserServlets(servletProducer, vaadinServlets);
+        }
+    }
+
+    private void registerUserServlets(BuildProducer<ServletBuildItem> servletProducer,
```

Comment:
the method still should be refactored: `for` doesn't have to be in the method, the inner block may be a method.

The method impl contains `for` inside `for` and `if` inside `for` and the `for` block is quite hard to read.